docs/azureai/azureai-openai-integration.md

```diff
@@ -1,138 +1,365 @@
+
+:::code language="bicep" source="../snippets/azure/AppHost/openai.module.bicep":::
+
+The preceding Bicep is a module that provisions an Azure Cognitive Services resource with the following defaults:
+
+- `location`: The location of the resource group.
+- `principalType`: The principal type of the Cognitive Services resource.
+- `principalId`: The principal ID of the Cognitive Services resource.
+- `openai`: The Cognitive Services account resource.
+  - `kind`: The kind of the resource, set to `OpenAI`.
+  - `properties`: The properties of the resource.
+    - `customSubDomainName`: The custom subdomain name for the resource, based on the unique string of the resource group ID.
+    - `publicNetworkAccess`: Set to `Enabled`.
+    - `disableLocalAuth`: Set to `true`.
+  - `sku`: The SKU of the resource, set to `S0`.
+- `openai_CognitiveServicesOpenAIContributor`: The Cognitive Services resource owner, based on the build-in `Azure Cognitive Services OpenAI Contributor` role. For more information, see [Azure Cognitive Services OpenAI Contributor](/azure/role-based-access-control/built-in-roles/ai-machine-learning#cognitive-services-openai-contributor).
+- `connectionString`: The connection string, containing the endpoint of the Cognitive Services resource.
+
+The generated Bicep is a starting point and can be customized to meet your specific requirements.
+
+### Customize provisioning infrastructure
+
+All .NET Aspire Azure resources are subclasses of the <xref:Aspire.Hosting.Azure.AzureProvisioningResource> type. This enables customization of the generated Bicep by providing a fluent API to configure the Azure resources—using the <xref:Aspire.Hosting.AzureProvisioningResourceExtensions.ConfigureInfrastructure``1(Aspire.Hosting.ApplicationModel.IResourceBuilder{``0},System.Action{Aspire.Hosting.Azure.AzureResourceInfrastructure})> API:
+
+:::code language="csharp" source="../snippets/azure/AppHost/Program.ConfigureOpenAIInfra.cs" id="configure":::
+
+The preceding code:
+
+- Chains a call to the <xref:Aspire.Hosting.AzureProvisioningResourceExtensions.ConfigureInfrastructure*> API:
+  - The `infra` parameter is an instance of the <xref:Aspire.Hosting.Azure.AzureResourceInfrastructure> type.
+  - The provisionable resources are retrieved by calling the <xref:Azure.Provisioning.Infrastructure.GetProvisionableResources> method.
+  - The single <xref:Azure.Provisioning.CognitiveServices.CognitiveServicesAccount> resource is retrieved.
+  - The <xref:Azure.Provisioning.CognitiveServices.CognitiveServicesAccount.Sku?displayProperty=nameWithType> property is assigned to a new instance of <xref:Azure.Provisioning.CognitiveServices.CognitiveServicesSku> with an `E0` name and <xref:Azure.Provisioning.CognitiveServices.CognitiveServicesSkuTier.Enterprise?displayProperty=nameWithType> tier.
+  - A tag is added to the Cognitive Services resource with a key of `ExampleKey` and a value of `Example value`.
+
+### Connect to an existing Azure OpenAI service
```

Comment:
Should this be updated to use the "AsExisting" functionality?

Reply:
I could update it, but there's a link to it after the code snippet - encouraging readers to learn more about those specific APIs.

Reply:
AddConnectionString is weakly typed - it is just a string - you don't know what it is pointing to: a storage account or a CosmosDB?

AsExisting is typed, so you know that it is Storage.

When using Azure Functions, you'd want AsExisting because when you `WithReference` it will set the right configuration settings for Functions.

My thinking is that we should guide users to use AsExisting where possible, it is the better API.

Reply:
Fixed